pkg/queue/forwarded_shim.go

```diff
@@ -0,0 +1,85 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"net/http"
+	"strings"
+)
+
+// ForwardedShimHandler attempts to shim a `forwarded` http header from the information
+// available in the `x-forwarded-*` headers. When available, each node in the `x-forwarded-for`
+// header is combined with the `x-forwarded-proto` and `x-forwarded-host` fields to construct
+// a `forwarded` header. The `x-forwarded-by` header is ignored entirely, since it cannot be
+// reliably combined with `x-forwarded-for`. No-op if a `forwarded` header is already present.
+// Note: IPv6 addresses are *not* supported.
+func ForwardedShimHandler(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Forwarded: by=<identifier>;for=<identifier>;host=<host>;proto=<http|https>
+		fwd := strings.TrimSpace(r.Header.Get(http.CanonicalHeaderKey("forwarded")))
+
+		// Don't add a shim if the header is already present
+		if fwd != "" {
+			h.ServeHTTP(w, r)
+			return
+		}
+
+		// X-Forwarded-For: <client>, <proxy1>, <proxy2>
+		xff := strings.TrimSpace(r.Header.Get(http.CanonicalHeaderKey("x-forwarded-for")))
+		// X-Forwarded-Proto: <protocol>
+		xfp := strings.TrimSpace(r.Header.Get(http.CanonicalHeaderKey("x-forwarded-proto")))
+		// X-Forwarded-Host: <host>
+		xfh := strings.TrimSpace(r.Header.Get(http.CanonicalHeaderKey("x-forwarded-host")))
+
+		// The forwarded header is a list of forwarded elements
+		elements := []string{}
+
+		// The x-forwarded-header consists of multiple nodes
+		nodes := strings.Split(xff, ",")
+
+		// The first element has a 'for', 'proto' and 'host' pair, as available
+		pairs := []string{}
+
+		if len(nodes) > 0 && nodes[0] != "" {
+			pairs = append(pairs, "for="+strings.TrimSpace(nodes[0]))
+		}
+		if xfh != "" {
+			pairs = append(pairs, "host="+xfh)
+		}
+		if xfp != "" {
+			pairs = append(pairs, "proto="+xfp)
+		}
+
+		// The pairs are joined with a semi-colon (;) into a single element
+		elements = append(elements, strings.Join(pairs, ";"))
```

Comment:
you can define elements right here and initialize it with capacity (len(nodes)).

Reply:
Not quite, see the test cases for x-forwarded-host and x-forwarded-proto without x-forwarded-for

Reply:
for performance?

Reply:
It also serves as a hint to the reader how are you going to populate it.
As far as performance, it's less crucial here, but if you're going to have a chain of more than 2 forwards you're going to end up with one additional allocation + copy at least.

Reply:
Also moving elements here makes it closer to the usage point. Now it's defined a page above and not used at all.

Reply:
I see where you're coming from. I think it's actually much more legible with a separate declaration and comment, but let me know if you feel strongly about this:

```go
		// The forwarded header is a list of forwarded elements
		elements := []string{}
```